<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#MIDS---w261-Machine-Learning-At-Scale" data-toc-modified-id="MIDS---w261-Machine-Learning-At-Scale-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>MIDS - w261 Machine Learning At Scale</a></span><ul class="toc-item"><li><span><a href="#Assignment---HW5" data-toc-modified-id="Assignment---HW5-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Assignment - HW5</a></span><ul class="toc-item"><li><span><a href="#INSTRUCTIONS-for-SUBMISSIONS" data-toc-modified-id="INSTRUCTIONS-for-SUBMISSIONS-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>INSTRUCTIONS for SUBMISSIONS</a></span></li></ul></li></ul></li><li><span><a href="#Optional-reading" data-toc-modified-id="Optional-reading-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Optional reading</a></span></li><li><span><a href="#HW-Problems" data-toc-modified-id="HW-Problems-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>HW Problems</a></span></li><li><span><a href="#HW5.0--data-warehouse;-star-schema" data-toc-modified-id="HW5.0--data-warehouse;-star-schema-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>HW5.0  data warehouse; star schema</a></span></li><li><span><a href="#HW5.1-Databases:-3NF;-denormalized" data-toc-modified-id="HW5.1-Databases:-3NF;-denormalized-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>HW5.1 Databases: 3NF; denormalized</a></span></li><li><span><a href="#HW5.2--Memory-backed-map-side" data-toc-modified-id="HW5.2--Memory-backed-map-side-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>HW5.2  Memory-backed map-side</a></span></li><li><span><a href="#HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join" data-toc-modified-id="HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>HW5.2.1 (OPTIONAL) Almost stateless reducer-side join</a></span></li><li><span><a href="#Pairwise-similarity----PHASE-1" data-toc-modified-id="Pairwise-similarity----PHASE-1-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Pairwise similarity  - PHASE 1</a></span><ul class="toc-item"><li><span><a href="#Run-Systems-tests-locally-on-small-datasets-(PHASE1)" data-toc-modified-id="Run-Systems-tests-locally-on-small-datasets-(PHASE1)-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Run Systems tests locally on small datasets (PHASE1)</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#1:-unit/systems-first-10-lines" data-toc-modified-id="1:-unit/systems-first-10-lines-8.1.0.1"><span class="toc-item-num">8.1.0.1&nbsp;&nbsp;</span>1: unit/systems first-10-lines</a></span></li><li><span><a href="#2:-unit/systems-atlas-boon" data-toc-modified-id="2:-unit/systems-atlas-boon-8.1.0.2"><span class="toc-item-num">8.1.0.2&nbsp;&nbsp;</span>2: unit/systems atlas-boon</a></span></li><li><span><a href="#3:-unit/systems-stripe-docs-test" data-toc-modified-id="3:-unit/systems-stripe-docs-test-8.1.0.3"><span class="toc-item-num">8.1.0.3&nbsp;&nbsp;</span>3: unit/systems stripe-docs-test</a></span></li></ul></li></ul></li><li><span><a href="#5.3-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.3-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>5.3 build stripes for all the test data sets - run the commands and insure that your output matches the output below</a></span></li><li><span><a href="#5.4-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.4-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>5.4 Build Inverted Index - run the commands and insure that your output matches the output below</a></span><ul class="toc-item"><li><span><a href="#Inverted-Index" data-toc-modified-id="Inverted-Index-8.3.1"><span class="toc-item-num">8.3.1&nbsp;&nbsp;</span>Inverted Index</a></span></li></ul></li><li><span><a href="#5.5-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.5-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>5.5 Calculate similarities - run the commands and insure that your output matches the output below</a></span><ul class="toc-item"><li><span><a href="#NOTE:-I-will-be-calculating-cosine-similarity,-which-means-I-need-to-re-write-my-inverted-index-MRJob" data-toc-modified-id="NOTE:-I-will-be-calculating-cosine-similarity,-which-means-I-need-to-re-write-my-inverted-index-MRJob-8.4.1"><span class="toc-item-num">8.4.1&nbsp;&nbsp;</span>NOTE: I will be calculating cosine similarity, which means I need to re-write my inverted index MRJob</a></span><ul class="toc-item"><li><span><a href="#NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities" data-toc-modified-id="NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities-8.4.1.1"><span class="toc-item-num">8.4.1.1&nbsp;&nbsp;</span>NOTE: you must run in hadoop mode to generate sorted similarities</a></span></li></ul></li><li><span><a href="#Pairwise-Similairity" data-toc-modified-id="Pairwise-Similairity-8.4.2"><span class="toc-item-num">8.4.2&nbsp;&nbsp;</span>Pairwise Similairity</a></span></li></ul></li></ul></li><li><span><a href="#===-END-OF-PHASE-1-===" data-toc-modified-id="===-END-OF-PHASE-1-===-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>=== END OF PHASE 1 ===</a></span></li></ul></div>

# MIDS - w261 Machine Learning At Scale
__Course Lead:__ Dr James G. Shanahan (__email__ Jimi via  James.Shanahan _AT_ gmail.com)

## Assignment - HW5


---
__Name:__  Stanimir Vichev   
__Class:__ MIDS w261 (Section 1, e.g., Group 1)     
__Email:__  stassyvichev@iSchool.Berkeley.edu 
__Week:__   5
    
__Due Time:__ HW is due the Thursday of the following week by 8AM (West coast time). I.e., Thursday, Feb 14, 2017 in the case of this homework. 

* __HW5 Phase 1__ 
This can be done on a local machine (with a unit test on the cloud such as Altiscale's PaaS or on AWS) and is due Thursday, Week 6 by 8AM (West coast time). It will primarily focus on building a unit/systems and for pairwise similarity calculations pipeline (for stripe documents)

* __HW5 Phase 2__ 
This will require the Altiscale cluster and will be due Thursday, Feb 22 by 8AM (West coast time). 
The focus of  HW5 Phase 2  will be to scale up the unit/systems tests to the Google 5 gram corpus. 



### INSTRUCTIONS for SUBMISSIONS 
Follow the instructions for submissions carefully.

Each student has a `HW-<user>` repository for all assignments.   

Push the following to your HW github repo into the master branch:
* Your local HW5 directory. Your repo file structure should look like this:

```
HW-<user>
    --HW3
       |__MIDS-W261-HW-03-<Student_id>.ipynb
       |__MIDS-W261-HW-03-<Student_id>.pdf
       |__some other hw3 file
    --HW4
       |__MIDS-W261-HW-04-<Student_id>.ipynb
       |__MIDS-W261-HW-04-<Student_id>.pdf
       |__some other hw4 file
    etc..
```    

# Optional reading
http://stanford.edu/~rezab/papers/disco.pdf   
https://terpconnect.umd.edu/~oard/pdf/acl08elsayed2.pdf

# HW Problems

# HW5.0  data warehouse; star schema

- What is a data warehouse? What is a Star schema? When is it used?

> A data warehouse is a set of tools and processes responsible for collecting, managing and serving the data that is needed for analysis by a business. It is usually a department within the company responsible for this, with a focus on data reliability, dependability and security, as well as analysis. 

> A star schema is a framework for a database architecture where different tables link to each other via primary and foreign keys, allowing information for a record to be stored separately in different tables in a normalized way. They are useful for records with a lot of attributes that can logically be split into attribute-specific tables for ease of storing, processing and logical analysis. Such tables heavily utilize joins to produce the complete record from the attributes. 

# HW5.1 Databases: 3NF; denormalized 

- In the database world What is 3NF? Does machine learning use data in 3NF? If so why? 
- In what form does ML consume data?
- Why would one use log files that are denormalized?

> A 3NF table in a database has several characteristics:

> 1. It has no cells with multiple values; Each row has a primary key (from 1NF)

> 2. The primary key contains multiple columns and non-key columns depend on the whole key ( from 2NF)

> 3. Each non-key attribute in a row does not depend on the entry in another non-key column.

> ML might have to work with 3NF data in situations where it uses information from a relational database, and RDBs are used extensively in business. As such, ML might have to pre-process the data in order to get all data for a record in one place. This is needed because ML might need all parameters associated with a record in order to do training on inference, since with ML we almost always want more data if we can have it rather than less. Log files are useful because they have all the data associated with a record in one row, which is quite convenient. 

In [106]:
# general imports
import os
import re
import sys
from os import environ
import numpy as np
import matplotlib.pyplot as plt
PATH  = environ['PATH']
# tell matplotlib not to open a new window
%matplotlib inline

# automatically reload modules 
%reload_ext autoreload
%autoreload 2

In [2]:
# print some configuration details for future replicability.
print 'Python Version: %s' % (sys.version.split('|')[0])
hdfs_conf = !hdfs getconf -confKey fs.defaultFS ### UNCOMMENT ON DOCKER
#hdfs_conf = !hdfs getconf -confKey fs.default.name ### UNCOMMENT ON ALTISCALE
print 'HDFS filesystem running at: \n\t %s' % (hdfs_conf[0])

Python Version: 2.7.14 
HDFS filesystem running at: 
	 hdfs://quickstart.cloudera:8020


In [3]:
# create an HDFS directory for this assignment
!hdfs dfs -mkdir hw51

In [3]:
HDFS_DIR = "/user/root/hw51" # eg. /user/root/hw3 
HOME_DIR = os.getcwd()

In [4]:
%%javascript
$('.math>span').css("border-left-color","transparent")

<IPython.core.display.Javascript object>

# HW5.2  Memory-backed map-side

Using MRJob, implement a hashside join (memory-backed map-side) for left, right and inner joins. Use the following tables for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table:

<pre>
transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA

Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

</pre>

**student task**


Please report the number of rows resulting from: (left = transactions, right = countries)

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right

__NOTE:__ For the Right and Inner joins, no reducer should be used. For the left outer join you can use a reducer.

__NOTE:__ If you need a refresher for MRJobs, please take a look at HW4.

> My prediction before I do this, just from looking at file:

> Left outer join: 9

> Right outer join: 9

> Inner join: 8

In [6]:
%%writefile Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

Writing Countries.dat


In [7]:
%%writefile transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA

Writing transactions.dat


In [54]:
%%writefile leftOuterJoin.py
#!/usr/bin/env python
#START STUDENT CODE44
from mrjob.job import MRJob
from mrjob.step import MRStep
import logging
class MRLeftOuterJoin(MRJob):
    
    SORT_VALUES = True
    JOBCONF = {"mapreduce.job.reduces": "1"}
    def __init__(self, *args, **kwargs):
        super(MRLeftOuterJoin, self).__init__(*args,**kwargs)
        self.countries = {}
        with open("Countries.dat") as file:
            for line in file.readlines():
                fields = line.split("|")
                self.countries[fields[1].replace("\n","")]=fields[0]
        
    def mapper(self, _, line):
        fields = line.split("|")
        yield fields[0], (fields[1], self.countries.get(fields[2]))
    
    def steps(self):
        return [
            MRStep(mapper = self.mapper)
        ]

if __name__ == "__main__":
    MRLeftOuterJoin.run()

Overwriting leftOuterJoin.py


In [5]:
from leftOuterJoin import MRLeftOuterJoin
# ,"-r", "hadoop"
mr_job = MRLeftOuterJoin(args=["transactions.dat","--file", "Countries.dat"])
count = 0
with mr_job.make_runner() as runner:
    runner.run()
    for line in runner.stream_output():
        count +=1
        print mr_job.parse_output_line(line)
print "Left outer line count: "+str(count)

(u'Alice Bob', [u'$10', u'United States'])
(u'Sam Sneed', [u'$1', u'Canada'])
(u'Jon Sneed', [u'$20', u'Canada'])
(u'Arnold Wesise', [u'$400', u'United Kingdom'])
(u'Henry Bob', [u'$2', u'United States'])
(u'Yo Yo Ma', [u'$2', u'Canada'])
(u'Jon York', [u'$44', u'Canada'])
(u'Alex Ball', [u'$5', u'United Kingdom'])
(u'Jim Davis', [u'$66', None])
Left outer line count: 9


In [64]:
%%writefile rightOuterJoin.py
#!/usr/bin/env python
#START STUDENT CODE44
from mrjob.job import MRJob
from mrjob.step import MRStep
import logging
class MRRightOuterJoin(MRJob):
    
    SORT_VALUES = True
    JOBCONF = {"mapreduce.job.reduces": "1"}
    def __init__(self, *args, **kwargs):
        super(MRRightOuterJoin, self).__init__(*args,**kwargs)
        self.countries = {}
        with open("Countries.dat") as file:
            for line in file.readlines():
                fields = line.split("|")
                self.countries[fields[1].replace("\n","")]=fields[0]
    
    def mapper(self, _, line):
        fields = line.split("|")
        if self.countries.get(fields[2]):
            yield self.countries.get(fields[2]), (fields[1], fields[0])
    
    def reducer_init(self):
        self.countries_done = []   
    
    def reducer_final(self):
        for key, value in self.countries.items():
            if value not in self.countries_done:
                yield value, None
    
    def reducer(self, key, value):
        if key in self.countries.values():
            self.countries_done.append(key)
        for v in value:
            yield key, v
            
    def steps(self):
        return [
            MRStep(mapper = self.mapper,
                   reducer = self.reducer,
                   reducer_init = self.reducer_init,
                   reducer_final = self.reducer_final)
        ]
if __name__ == "__main__":
    MRRightOuterJoin.run()

Overwriting rightOuterJoin.py


In [6]:
from rightOuterJoin import MRRightOuterJoin
# ,"-r", "hadoop"
mr_job = MRRightOuterJoin(args=["transactions.dat","--file", "Countries.dat"])
count = 0
with mr_job.make_runner() as runner:
    runner.run()
    for line in runner.stream_output():
        count +=1
        print mr_job.parse_output_line(line)
print "Right outer line count: "+str(count)

(u'Canada', [u'$1', u'Sam Sneed'])
(u'Canada', [u'$2', u'Yo Yo Ma'])
(u'Canada', [u'$20', u'Jon Sneed'])
(u'Canada', [u'$44', u'Jon York'])
(u'United Kingdom', [u'$400', u'Arnold Wesise'])
(u'United Kingdom', [u'$5', u'Alex Ball'])
(u'United States', [u'$10', u'Alice Bob'])
(u'United States', [u'$2', u'Henry Bob'])
(u'Italy', None)
Right outer line count: 9


In [67]:
%%writefile innerJoin.py
#!/usr/bin/env python
#START STUDENT CODE44
from mrjob.job import MRJob
from mrjob.step import MRStep
import logging
class MRInnerJoin(MRJob):
    
    SORT_VALUES = True
    JOBCONF = {"mapreduce.job.reduces": "1"}
    def __init__(self, *args, **kwargs):
        super(MRInnerJoin, self).__init__(*args,**kwargs)
        self.countries = {}
        with open("Countries.dat") as file:
            for line in file.readlines():
                fields = line.split("|")
                self.countries[fields[1].replace("\n","")]=fields[0]
        
    def mapper(self, _, line):
        fields = line.split("|")
        if self.countries.get(fields[2]):
            yield fields[0], (fields[1], self.countries.get(fields[2]))
    
    def steps(self):
        return [
            MRStep(mapper = self.mapper)
        ]

if __name__ == "__main__":
    MRInnerJoin.run()

Writing innerJoin.py


In [7]:
from innerJoin import MRInnerJoin
# ,"-r", "hadoop"
mr_job = MRInnerJoin(args=["transactions.dat","--file", "Countries.dat"])
count = 0
with mr_job.make_runner() as runner:
    runner.run()
    for line in runner.stream_output():
        count +=1
        print mr_job.parse_output_line(line)
print "Inner line count: "+str(count)

(u'Alice Bob', [u'$10', u'United States'])
(u'Sam Sneed', [u'$1', u'Canada'])
(u'Jon Sneed', [u'$20', u'Canada'])
(u'Arnold Wesise', [u'$400', u'United Kingdom'])
(u'Henry Bob', [u'$2', u'United States'])
(u'Yo Yo Ma', [u'$2', u'Canada'])
(u'Jon York', [u'$44', u'Canada'])
(u'Alex Ball', [u'$5', u'United Kingdom'])
Inner line count: 8


# HW5.2.1 (OPTIONAL) Almost stateless reducer-side join

The following MRJob code, implements a reduce-side join for an inner join. The reducer is almost stateless, i.e., uses as little memory as possible. Use the tables from HW5.2 for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table perform. Perform  an left, right, inner joins using the code provided below and report the number of rows resulting from:

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right

Again make smart decisions about which table should be the left table (i.e., crosscheck the code). 

__Some notes on the code__ 
Here, the mapper receives its set of input splits either from the transaction table or from the countries table and makes the appropriate transformations: splitting the line into fields, and emitting a key/value. The key is the join key - in this case, the country code field of both sets of records. The mapper knows which file and type of record it is receiving based on the length of the fields. The records it emits contain the join field as the key, which acts as the partitioning key; We use the SORT_VALUES option, which ensures the values are sorted as well. Then, we employ a trick to ensure that for each join key, country records are seen always before transaction records. We achieve this by adding an arbitrary key to the front of the value: 'A' for countries, 'B' for customers. This makes countries sort before customers for each and every join/partition key. After that trick, the join is simply a matter of storing countries ('A' records) and crossing this array with each customer record.

In [ ]:
import sys, os, re
from mrjob.job import MRJob

class MRJoin(MRJob):

  # Performs secondary sort
  SORT_VALUES = True

  def mapper(self, _, line):
    splits = line.rstrip("\n").split("|")

    if len(splits) == 2: # country data
      symbol = 'A' # make country sort before transaction data
      country2digit = splits[1]
      yield country2digit, [symbol, splits]
    else: # person data
      symbol = 'B'
      country2digit = splits[2]
      yield country2digit, [symbol, splits]

  def reducer(self, key, values):
    countries = [] # should come first, as they are sorted on artificia key 'A'
    for value in values:
      if value[0] == 'A':
        countries.append(value)
      if value[0] == 'B':
        for country in countries:
          yield key, country[1:] + value[1:]

if __name__ == '__main__':
  MRJoin.run()

# Pairwise similarity  - PHASE 1

In this part of the assignment we will focus on developing methods for detecting synonyms, using the Google 5-grams dataset. To accomplish this you must script three main tasks using MRJob:


- (1) Using the systems tests data sets, write mrjob code to build the stripes
- (2) Write mrjob code to build an inverted index from the stripes
- (3) Using two (symmetric) comparison methods of your choice (e.g., correlations, distances, similarities), pairwise compare all stripes (vectors), and output to a file.   


For this task you will have to determine a method of comparison.
Here are a few that you might consider:

 - Jaccard
 - Cosine similarity
 - Spearman correlation
 - Euclidean distance
 - Taxicab (Manhattan) distance
 - Shortest path graph distance (a graph, because our data is symmetric!)
 - Pearson correlation
 - Kendall correlation
 
 ...
 
 (for more details: http://stanford.edu/~rezab/papers/disco.pdf pg5, table 1)

However, be cautioned that some comparison methods are more difficult to parallelize than others, and do not perform more associations than is necessary, since your choice of association will be symmetric.

Please use the inverted index (discussed in live session #5) based pattern to compute the pairwise (term-by-term) similarity matrix. 

## Run Systems tests locally on small datasets (PHASE1)

Complete 5.3-5.5 and systems test using the below test datasets. Phase 2 will focus on the entire Ngram dataset.

To help you through these tasks please verify that your code gives the results below (for stripes, inverted index, and pairwise similarities).

Test datasets:

* googlebooks-eng-all-5gram-20090715-0-filtered.txt [see below]
* atlas-boon-test [see below]
* stripe-docs-test [see below]


A large subset of the Google n-grams dataset   
https://aws.amazon.com/datasets/google-books-ngrams/

which we have placed in a bucket/folder on Dropbox:   
https://www.dropbox.com/sh/tmqpc4o0xswhkvz/AACUifrl6wrMrlK6a3X3lZ9Ea?dl=0 

You can download from Dropbox using wget:   
`wget -O googlebooks-eng-all-5gram-20090715-0-filtered.txt https://www.dropbox.com/sh/tmqpc4o0xswhkvz/AACr50woxiBWoaiiLmnwduX8a/googlebooks-eng-all-5gram-20090715-0-filtered.txt?dl=0`


In particular, this bucket contains (~200) files (10Meg each) in the format:

	(ngram) \t (count) \t (pages_count) \t (books_count)

The next cell shows the first 10 lines of the googlebooks-eng-all-5gram-20090715-0-filtered.txt file.


__DISCLAIMER__: Each record is already a 5-gram. In real life, we would calculate the stripes cooccurrence data from the raw text by windowing over the raw text and not from the 5-gram preprocessed data (as we are doing here).  Calculatating pairs on this 5-gram is a little corrupt as we will be double counting cooccurences. Having said that this exercise can still pull out some simialr terms. 

#### 1: unit/systems first-10-lines

In [9]:
%%writefile googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt
A BILL FOR ESTABLISHING RELIGIOUS	59	59	54
A Biography of General George	92	90	74
A Case Study in Government	102	102	78
A Case Study of Female	447	447	327
A Case Study of Limited	55	55	43
A Child's Christmas in Wales	1099	1061	866
A Circumstantial Narrative of the	62	62	50
A City by the Sea	62	60	49
A Collection of Fairy Tales	123	117	80
A Collection of Forms of	116	103	82

Writing googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt


#### 2: unit/systems atlas-boon

In [10]:
%%writefile atlas-boon-systems-test.txt
atlas boon	50	50	50
boon cava dipped	10	10	10
atlas dipped	15	15	15

Writing atlas-boon-systems-test.txt


#### 3: unit/systems stripe-docs-test

Three terms, A,B,C and their corresponding stripe-docs of co-occurring terms

- DocA {X:20, Y:30, Z:5}
- DocB {X:100, Y:20}
- DocC {M:5, N:20, Z:5}

## 5.3 build stripes for all the test data sets - run the commands and insure that your output matches the output below

* __Note__: Your results may not match 100%.

In [13]:
%%writefile buildStripes.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import re
import mrjob
import json
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRbuildStripes(MRJob):
    SORT_VALUES = True
    JOBCONF = {"mapreduce.job.reduces": "1"}
    
    def mapper(self, _, line):
        fields = line.lower().strip("\n").split("\t")
        words = fields[0].split(" ")
        occurrence_count = int(fields[1])
        count = 1
        for word in words:
            for i in xrange(count, len(words)):
                yield word, (words[i], occurrence_count)
                yield words[i], (word, occurrence_count)
            count += 1
    
    def reducer_init(self):
        print
        
    def reducer(self, word, occurrence_counts):
        stripe = {}
        for other_word, occurrence_count in occurrence_counts:
            stripe[other_word] = stripe.get(other_word,0)+occurrence_count
        yield word, stripe

if __name__ == '__main__':
    MRbuildStripes.run()

Writing buildStripes.py


In [8]:
###########################################################################
# Make Stripes from ngrams for systems test 1
###########################################################################

!hdfs dfs -rm --recursive systems_test_stripes_1
!python buildStripes.py -r local googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt > systems_test_stripes_1 

-rm: Illegal option --recursive
Usage: hadoop fs [generic options] -rm [-f] [-r|-R] [-skipTrash] <src> ...
No configs found; falling back on auto-configuration
Creating temp directory /tmp/buildStripes.root.20180211.101746.383545
Running step 1 of 1...
Streaming final output from /tmp/buildStripes.root.20180211.101746.383545/output...
Removing temp directory /tmp/buildStripes.root.20180211.101746.383545...


In [9]:
!cat systems_test_stripes_1


"a"	{"limited":55,"sea":62,"general":92,"female":447,"in":1201,"religious":59,"george":92,"biography":92,"city":62,"for":59,"tales":123,"government":102,"the":124,"forms":116,"wales":1099,"christmas":1099,"child's":1099,"collection":239,"by":62,"case":604,"circumstantial":62,"of":1011,"study":604,"bill":59,"establishing":59,"narrative":62,"fairy":123}
"bill"	{"a":59,"religious":59,"for":59,"establishing":59}
"biography"	{"a":92,"of":92,"george":92,"general":92}
"by"	{"a":62,"city":62,"the":62,"sea":62}
"case"	{"a":604,"limited":55,"government":102,"of":502,"study":604,"female":447,"in":102}
"child's"	{"a":1099,"wales":1099,"christmas":1099,"in":1099}
"christmas"	{"a":1099,"wales":1099,"in":1099,"child's":1099}
"circumstantial"	{"a":62,"of":62,"the":62,"narrative":62}
"city"	{"a":62,"the":62,"by":62,"sea":62}
"collection"	{"a":239,"forms":116,"fairy":123,"tales":123,"of":355}
"establishing"	{"a":59,"bill":59,"religious":59,"for":59}
"fairy"	{"a":123,"of":123,"tales":123,"collection":12

<pre>
"a"	{"limited": 55, "female": 447, "general": 92, "sea": 62, "in": 1201, "religious": 59, "george": 92, "biography": 92, "city": 62, "for": 59, "tales": 123, "child's": 1099, "by": 62, "forms": 116, "wales": 1099, "christmas": 1099, "government": 102, "collection": 239, "fairy": 123, "case": 604, "circumstantial": 62, "of": 1011, "study": 604, "bill": 59, "establishing": 59, "narrative": 62, "the": 124}
"bill"	{"a": 59, "religious": 59, "for": 59, "establishing": 59}
"biography"	{"a": 92, "of": 92, "george": 92, "general": 92}
"by"	{"a": 62, "city": 62, "the": 62, "sea": 62}
"case"	{"a": 604, "limited": 55, "government": 102, "of": 502, "study": 604, "female": 447, "in": 102}
"child's"	{"a": 1099, "wales": 1099, "christmas": 1099, "in": 1099}
"christmas"	{"a": 1099, "wales": 1099, "in": 1099, "child's": 1099}
"circumstantial"	{"a": 62, "of": 62, "the": 62, "narrative": 62}
"city"	{"a": 62, "the": 62, "by": 62, "sea": 62}
"collection"	{"a": 239, "forms": 116, "fairy": 123, "tales": 123, "of": 355}
"establishing"	{"a": 59, "bill": 59, "religious": 59, "for": 59}
"fairy"	{"a": 123, "of": 123, "tales": 123, "collection": 123}
"female"	{"a": 447, "case": 447, "study": 447, "of": 447}
"for"	{"a": 59, "bill": 59, "religious": 59, "establishing": 59}
"forms"	{"a": 116, "of": 232, "collection": 116}
"general"	{"a": 92, "of": 92, "george": 92, "biography": 92}
"george"	{"a": 92, "of": 92, "biography": 92, "general": 92}
"government"	{"a": 102, "case": 102, "study": 102, "in": 102}
"in"	{"a": 1201, "case": 102, "government": 102, "study": 102, "child's": 1099, "wales": 1099, "christmas": 1099}
"limited"	{"a": 55, "case": 55, "study": 55, "of": 55}
"narrative"	{"a": 62, "of": 62, "the": 62, "circumstantial": 62}
"of"	{"a": 1011, "case": 502, "circumstantial": 62, "george": 92, "limited": 55, "tales": 123, "collection": 355, "general": 92, "forms": 232, "female": 447, "narrative": 62, "study": 502, "fairy": 123, "the": 62, "biography": 92}
"religious"	{"a": 59, "bill": 59, "for": 59, "establishing": 59}
"sea"	{"a": 62, "city": 62, "the": 62, "by": 62}
"study"	{"a": 604, "case": 604, "limited": 55, "government": 102, "of": 502, "female": 447, "in": 102}
"tales"	{"a": 123, "of": 123, "fairy": 123, "collection": 123}
"the"	{"a": 124, "city": 62, "circumstantial": 62, "of": 62, "sea": 62, "narrative": 62, "by": 62}
"wales"	{"a": 1099, "in": 1099, "christmas": 1099, "child's": 1099}

</pre>

In [10]:
###########################################################################
# Make Stripes from ngrams for systems test 2
###########################################################################

!hdfs dfs -rm --recursive systems_test_stripes_2
!python buildStripes.py -r local atlas-boon-systems-test.txt > systems_test_stripes_2

-rm: Illegal option --recursive
Usage: hadoop fs [generic options] -rm [-f] [-r|-R] [-skipTrash] <src> ...
No configs found; falling back on auto-configuration
Creating temp directory /tmp/buildStripes.root.20180211.101756.853750
Running step 1 of 1...
Streaming final output from /tmp/buildStripes.root.20180211.101756.853750/output...
Removing temp directory /tmp/buildStripes.root.20180211.101756.853750...


In [11]:
!cat systems_test_stripes_2


"atlas"	{"dipped":15,"boon":50}
"boon"	{"atlas":50,"dipped":10,"cava":10}
"cava"	{"dipped":10,"boon":10}
"dipped"	{"atlas":15,"boon":10,"cava":10}


<pre>
"atlas"   {"dipped": 15, "boon": 50}   
"boon"    {"atlas": 50, "dipped": 10, "cava": 10}   
"cava"    {"dipped": 10, "boon": 10} 
"dipped"  {"atlas": 15, "boon": 10, "cava": 10}
</pre>

In [60]:
########################################################################
# Stripes for systems test 3 (given, no need to build stripes)
########################################################################

with open("systems_test_stripes_3", "w") as f:
    f.writelines([
        '"DocA"\t{"X":20,"Y":30,"Z":5}\n',
        '"DocB"\t{"X":100,"Y":20}\n',  
        '"DocC"\t{"M":5,"N":20,"Z":5,"Y":1}\n'
    ])
!cat systems_test_stripes_3   

"DocA"	{"X":20,"Y":30,"Z":5}
"DocB"	{"X":100,"Y":20}
"DocC"	{"M":5,"N":20,"Z":5,"Y":1}


## 5.4 Build Inverted Index - run the commands and insure that your output matches the output below

* You might find this [paper](https://terpconnect.umd.edu/~oard/pdf/acl08elsayed2.pdf) useful.

In [21]:
%%writefile invertedIndex.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-


from __future__ import division
import collections
import re
import json
import math
import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRinvertedIndex(MRJob):
    
    def mapper(self, _, line):
        if line:
            word,stripe = line.split()
            stripe=json.loads(stripe)
            len_dict = len(stripe)
            for key, _ in stripe.items():
                yield key, (word.strip("\""), len_dict)
    
    def reducer(self, key, values):
        yield key, values
        
if __name__ == '__main__':
    MRinvertedIndex.run() 

Overwriting invertedIndex.py


In [24]:
!python invertedIndex.py -r local systems_test_stripes_1 > systems_test_index_1

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180211.102158.103698
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180211.102158.103698/output...
Removing temp directory /tmp/invertedIndex.root.20180211.102158.103698...


In [25]:
!python invertedIndex.py -r local systems_test_stripes_2 > systems_test_index_2

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180211.102202.953187
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180211.102202.953187/output...
Removing temp directory /tmp/invertedIndex.root.20180211.102202.953187...


In [26]:
!python invertedIndex.py -r local systems_test_stripes_3 > systems_test_index_3

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180211.102207.028985
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180211.102207.028985/output...
Removing temp directory /tmp/invertedIndex.root.20180211.102207.028985...


In [27]:
##########################################################
# Pretty print systems tests for generating Inverted Index
##########################################################

import json

for i in range(1,4):
    print "—"*100
    print "Systems test ",i," - Inverted Index"
    print "—"*100  
    with open("systems_test_index_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            word,stripe = line.split("\t")
            stripe = json.loads(stripe)
            stripe.extend([["",""] for _ in xrange(3 - len(stripe))])
            print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
                (word), stripe[0][0]+" "+str(stripe[0][1]), stripe[1][0]+" "+str(stripe[1][1]), stripe[2][0]+" "+str(stripe[2][1]))
        


————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |          bill 4 |     biography 4 |            by 4
          "bill" |            a 27 |  establishing 4 |           for 4
     "biography" |            a 27 |       general 4 |        george 4
            "by" |            a 27 |          city 4 |           sea 4
          "case" |            a 27 |        female 4 |    government 4
       "child's" |            a 27 |     christmas 4 |            in 7
     "christmas" |            a 27 |       child's 4 |            in 7
"circumstantial" |            a 27 |     narrative 4 |           of 16
          "city" |            a 27 |            by 4 |           sea 4
    "collection" |            a 27 |         fairy 4 |         forms 3
  "establishing" |            a 27 |          bill 4 |

### Inverted Index

In [ ]:
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |          bill 4 |     biography 4 |            by 4
          "bill" |            a 27 |  establishing 4 |           for 4
     "biography" |            a 27 |       general 4 |        george 4
            "by" |            a 27 |          city 4 |           sea 4
          "case" |            a 27 |        female 4 |    government 4
       "child's" |            a 27 |     christmas 4 |            in 7
     "christmas" |            a 27 |       child's 4 |            in 7
"circumstantial" |            a 27 |     narrative 4 |           of 15
          "city" |            a 27 |            by 4 |           sea 4
    "collection" |            a 27 |         fairy 4 |         forms 3
  "establishing" |            a 27 |          bill 4 |           for 4
         "fairy" |            a 27 |    collection 5 |           of 15
        "female" |            a 27 |          case 7 |           of 15
           "for" |            a 27 |          bill 4 |  establishing 4
         "forms" |            a 27 |    collection 5 |           of 15
       "general" |            a 27 |     biography 4 |        george 4
        "george" |            a 27 |     biography 4 |       general 4
    "government" |            a 27 |          case 7 |            in 7
            "in" |            a 27 |          case 7 |       child's 4
       "limited" |            a 27 |          case 7 |           of 15
     "narrative" |            a 27 |circumstantial 4 |           of 15
            "of" |            a 27 |     biography 4 |          case 7
     "religious" |            a 27 |          bill 4 |  establishing 4
           "sea" |            a 27 |            by 4 |          city 4
         "study" |            a 27 |          case 7 |        female 4
         "tales" |            a 27 |    collection 5 |         fairy 4
           "the" |            a 27 |            by 4 |circumstantial 4
         "wales" |            a 27 |       child's 4 |     christmas 4
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
         "atlas" |          boon 3 |        dipped 3 |                
          "boon" |         atlas 2 |          cava 2 |        dipped 3
          "cava" |          boon 3 |        dipped 3 |                
        "dipped" |         atlas 2 |          boon 3 |          cava 2
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "M" |          DocC 4 |                 |                
             "N" |          DocC 4 |                 |                
             "X" |          DocA 3 |          DocB 2 |                
             "Y" |          DocA 3 |          DocB 2 |          DocC 4
             "Z" |          DocA 3 |          DocC 4 |                


## 5.5 Calculate similarities - run the commands and insure that your output matches the output below

### NOTE: I will be calculating cosine similarity, which means I need to re-write my inverted index MRJob

In [59]:
%%writefile invertedIndexCosine.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-


from __future__ import division
import collections
import re
import json
import math
import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRinvertedIndexCosine(MRJob):
    
    def mapper(self, _, line):
        if line:
            word,stripe = line.split()
            stripe=json.loads(stripe)
            stripe_length = 1/math.sqrt(len(stripe))
            for key, _ in stripe.items():
                yield key, (word.strip("\""), stripe_length)
    
    def reducer(self, key, values):
        yield key, values
        
if __name__ == '__main__':
    MRinvertedIndexCosine.run() 

Overwriting invertedIndexCosine.py


In [61]:
!python invertedIndexCosine.py -r local systems_test_stripes_1 > systems_test_index_cosine_1

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndexCosine.root.20180211.115513.567171
Running step 1 of 1...
Streaming final output from /tmp/invertedIndexCosine.root.20180211.115513.567171/output...
Removing temp directory /tmp/invertedIndexCosine.root.20180211.115513.567171...


In [62]:
!python invertedIndexCosine.py -r local systems_test_stripes_2 > systems_test_index_cosine_2

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndexCosine.root.20180211.115516.616693
Running step 1 of 1...
Streaming final output from /tmp/invertedIndexCosine.root.20180211.115516.616693/output...
Removing temp directory /tmp/invertedIndexCosine.root.20180211.115516.616693...


In [63]:
!python invertedIndexCosine.py -r local systems_test_stripes_3 > systems_test_index_cosine_3

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndexCosine.root.20180211.115519.710130
Running step 1 of 1...
Streaming final output from /tmp/invertedIndexCosine.root.20180211.115519.710130/output...
Removing temp directory /tmp/invertedIndexCosine.root.20180211.115519.710130...


In [64]:
##########################################################
# Pretty print systems tests for generating Inverted Index Cosine
##########################################################

import json

for i in range(1,4):
    print "—"*100
    print "Systems test ",i," - Inverted Index"
    print "—"*100  
    with open("systems_test_index_cosine_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            word,stripe = line.split("\t")
            stripe = json.loads(stripe)
            stripe.extend([["",""] for _ in xrange(3 - len(stripe))])
            print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
                (word), stripe[0][0]+" "+str(stripe[0][1]), stripe[1][0]+" "+str(stripe[1][1]), stripe[2][0]+" "+str(stripe[2][1]))
        


————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |        bill 0.5 |   biography 0.5 |          by 0.5
          "bill" |  a 0.1924500897 |establishing 0.5 |         for 0.5
     "biography" |  a 0.1924500897 |     general 0.5 |      george 0.5
            "by" |  a 0.1924500897 |        city 0.5 |         sea 0.5
          "case" |  a 0.1924500897 |      female 0.5 |  government 0.5
       "child's" |  a 0.1924500897 |   christmas 0.5 |  in 0.377964473
     "christmas" |  a 0.1924500897 |     child's 0.5 |  in 0.377964473
"circumstantial" |  a 0.1924500897 |   narrative 0.5 |         of 0.25
          "city" |  a 0.1924500897 |          by 0.5 |         sea 0.5
    "collection" |  a 0.1924500897 |       fairy 0.5 |forms 0.5773502692
  "establishing" |  a 0.1924500897 |        bill 0.5

In [125]:
%%writefile similarity.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-
import json
import math
import mrjob
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRSimilarity(MRJob):
    SORT_VALUES = True
    JOBCONF = {"mapreduce.job.reduces": "1"}
    
    def mapper_first(self, _, line):
        word, stripe = line.split("\t")
        stripe = json.loads(stripe)
        if len(stripe)>1:
            count = 1
            for s in stripe:
                for i in xrange(count, len(stripe)):
                    yield (s[0], stripe[i][0]), s[1]*stripe[i][1]*1.0
                count += 1
    
    def reducer_first(self, pair, partSims):
        sum_sims = sum(partSims)
        yield sum_sims, (pair, sum_sims)
        
    def reducer_second(self, key, value):
        yield float(key), list(value)[0]
        
    def steps(self):
        return [
            MRStep(
                mapper = self.mapper_first,
                reducer = self.reducer_first
            ),
            MRStep(
                reducer = self.reducer_second,
                jobconf={
                        "mapreduce.job.reduces": "1",
                        "stream.num.map.output.key.fields": 1,
                        "mapreduce.job.output.key.comparator.class" : "org.apache.hadoop.mapred.lib.KeyFieldBasedComparator",
                        "mapreduce.partition.keycomparator.options":"-k1,1nr",
                        "mapred.num.key.comparator.options":"-k1,1nr",
                        "mapred.text.key.comparator.options": "-k1,1nr",
                        "SORT_VALUES":True
                   }
            )
        ]
        
if __name__ == '__main__':
    MRSimilarity.run()

Overwriting similarity.py


#### NOTE: you must run in hadoop mode to generate sorted similarities

In [126]:
!python similarity.py -r hadoop systems_test_index_cosine_1 > systems_test_similarities_1

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180211.125320.735958
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180211.125320.735958/files/...
Running step 1 of 2...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob6593545834351242203.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518344134162_0015
  Submitted application application_1518344134162_0015
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [127]:
!python similarity.py -r hadoop systems_test_index_cosine_2 > systems_test_similarities_2

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180211.125444.655159
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180211.125444.655159/files/...
Running step 1 of 2...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob3818410674907357171.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518344134162_0017
  Submitted application application_1518344134162_0017
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [128]:
!python similarity.py -r hadoop systems_test_index_cosine_3 > systems_test_similarities_3

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180211.125556.281154
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180211.125556.281154/files/...
Running step 1 of 2...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob1091797265403280663.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518344134162_0019
  Submitted application application_1518344134162_0019
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [129]:
############################################
# Pretty print systems tests
# Note: adjust print formatting if you need to
############################################

import json
for i in range(1,4):
    print '—'*110
    print "Systems test ",i," - Similarity measures"
    print '—'*110
    print "{0:>15} |{1:>15} |{2:>15}".format(
          "average", "pair", "cosine")
    print '-'*110

    with open("systems_test_similarities_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            avg,stripe = line.split("\t")
            stripe = json.loads(stripe)

            print "{0:>15f} |{1:>15} |{2:>15f} ".format(
              float(avg), stripe[0], float(stripe[1]))
 

——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
        average |           pair |         cosine
--------------------------------------------------------------------------------------------------------------
       1.000000 |[u'female', u'limited'] |       1.000000 
       0.866025 |[u'forms', u'tales'] |       0.866025 
       0.857143 |[u'case', u'study'] |       0.857143 
       0.750000 |[u'bill', u'establishing'] |       0.750000 
       0.721688 |  [u'a', u'of'] |       0.721688 
       0.670820 |[u'collection', u'tales'] |       0.670820 
       0.577350 |[u'forms', u'narrative'] |       0.577350 
       0.566947 |[u'case', u'limited'] |       0.566947 
       0.559017 |[u'collection', u'of'] |       0.559017 
       0.516398 |[u'collection', u'forms'] |       0.516398

### Pairwise Similairity 

In [ ]:
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |    female - limited |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.868292 |       fairy - forms |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.868292 |       forms - tales |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.830357 |        case - study |       0.857143 |       0.750000 |       0.857143 |       0.857143
  0.712500 | bill - establishing |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |   christmas - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |circumstantial - narrative |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |            by - sea |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |           by - city |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |     child's - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |  biography - george |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 | child's - christmas |       0.750000 |       0.600000 |       0.750000 |       0.750000
  ...
  
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |        atlas - cava |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.625000 |       boon - dipped |       0.666667 |       0.500000 |       0.666667 |       0.666667
  0.389562 |       cava - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |         boon - cava |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |      atlas - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |        atlas - boon |       0.408248 |       0.250000 |       0.500000 |       0.400000
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  0.820791 |         DocA - DocB |       0.816497 |       0.666667 |       1.000000 |       0.800000
  0.553861 |         DocA - DocC |       0.577350 |       0.400000 |       0.666667 |       0.571429
  0.346722 |         DocB - DocC |       0.353553 |       0.200000 |       0.500000 |       0.333333


# === END OF PHASE 1 ===